### ISSUE TITLE
InvalidGrantError: (invalid_grant) Token has been expired or revoked.

### ISSUE THREAD

**Issue by lutzky** [gilesknap/gphotos-sync#283]
I have been running gphotos-sync on a nightly cron job, and it last succeeded 2021-01-11 at 1AM. Ever since, it has been failing with this issue (both with the snap version and the pip version; this error is from pip version 2.14.2):

```
01-16 20:54:54 WARNING  gphotos-sync 2.14.2 2021-01-16 20:54:54.785955 
01-16 20:54:54 ERROR    
Process failed. 
Traceback (most recent call last):
  File "/home/ohad/.local/share/virtualenvs/photos-aX3072vj/lib/python3.8/site-packages/requests_oauthlib/oauth2_session.py", line 477, in request
    url, headers, data = self._client.add_token(
  File "/home/ohad/.local/share/virtualenvs/photos-aX3072vj/lib/python3.8/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 198, in add_token
    raise TokenExpiredError()
oauthlib.oauth2.rfc6749.errors.TokenExpiredError: (token_expired) 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ohad/.local/share/virtualenvs/photos-aX3072vj/lib/python3.8/site-packages/gphotos/Main.py", line 417, in main
    self.setup(args, db_path)
  File "/home/ohad/.local/share/virtualenvs/photos-aX3072vj/lib/python3.8/site-packages/gphotos/Main.py", line 309, in setup
    self.google_photos_client = RestClient(photos_api_url, self.auth.session)
  File "/home/ohad/.local/share/virtualenvs/photos-aX3072vj/lib/python3.8/site-packages/gphotos/restclient.py", line 38, in __init__
    service_document = self.auth_session.get(api_url).json()
  File "/home/ohad/.local/share/virtualenvs/photos-aX3072vj/lib/python3.8/site-packages/requests/sessions.py", line 555, in get
    return self.request('GET', url, **kwargs)
  File "/home/ohad/.local/share/virtualenvs/photos-aX3072vj/lib/python3.8/site-packages/requests_oauthlib/oauth2_session.py", line 496, in request
    token = self.refresh_token(
  File "/home/ohad/.local/share/virtualenvs/photos-aX3072vj/lib/python3.8/site-packages/requests_oauthlib/oauth2_session.py", line 446, in refresh_token
    self.token = self._client.parse_request_body_response(r.text, scope=self.scope)
  File "/home/ohad/.local/share/virtualenvs/photos-aX3072vj/lib/python3.8/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 421, in parse_request_body_response
    self.token = parse_token_response(body, scope=scope)
  File "/home/ohad/.local/share/virtualenvs/photos-aX3072vj/lib/python3.8/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 431, in parse_token_response
    validate_token_parameters(params)
  File "/home/ohad/.local/share/virtualenvs/photos-aX3072vj/lib/python3.8/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 438, in validate_token_parameters
    raise_from_error(params.get('error'), params)
  File "/home/ohad/.local/share/virtualenvs/photos-aX3072vj/lib/python3.8/site-packages/oauthlib/oauth2/rfc6749/errors.py", line 405, in raise_from_error
    raise cls(**kwargs)
oauthlib.oauth2.rfc6749.errors.InvalidGrantError: (invalid_grant) Token has been expired or revoked.
01-16 20:54:54 WARNING  Done. 
```

I haven't revoked the token nor does it have an expiry date. I've created a second oauth key, and using it I get the same result.

**Comment by lutzky** [comment]
Running this with a blank directory (but same `client_secret.json`) seems to work around the issue - i.e. it sends me to the authorize URL, and is currently "Indexing Google Photos Files". Not a fantastic workaround though.

**Comment by gilesknap** [comment]
Hi, occasionally the authorization token refresh fails and you will get his
(its not the oauth secret key so recreating that does not help)

if you run gphotos sync with --new-token it will ask you for your login
details again and fix the issue (--new-token is from memory and I
cant check right now - use --help if I have got the option name wrong)

**Comment by lutzky** [comment]
Yep, just found that in `--help` and it did indeed solve the problem. I think this could be a great improvement to the error message, recommending re-running with that flag.

**Comment by gilesknap** [comment]
yep, the message is not mine but there may be an opportunity to trap that
exception and make a better message. I'll put it on the todo list.

**Comment by ve6rah** [comment]
It's more than just the wording of the error message. Using gphotos-sync nightly this expiry thing hits me about once a week, and the only way to fix it is to do the --new-token thing manually. This is way too often. I have other programs that use Google's OAuth2 and none of them ever need me to re-do the authentication, so why does gphotos-sync need it weekly?

**Comment by gilesknap** [comment]
@ve6rah I'm sorry, but I cant explain that. I have only seen failed token refresh once in the whole time I have used gphotos-sync. In fact my CI runs system tests against a real Google Account every time I check in changes. That CI has run for more than 4 years and its token has always refreshed automatically.

I wonder if there is a subtlety in the way that you originally generated the secret that is causing this? I vaguely recall that there are refresh options in secret's configuration. I can only suggest generating your secret again to see if that helps. See https://docs.google.com/document/d/1ck1679H8ifmZ_4eVbDeD_-jezIcZ-j6MlaNaeQiz7y0/edit?usp=sharing

**Comment by ve6rah** [comment]
I tried following your document, and run into a gap in it's explanation. It doesn't go through which scopes to add. The screenshots you show are of an external app (mine previously was "testing" so I figured that might be the issue) but it also shows not requiring verification. When I allow read access to google photos, it shows as needing verification for an external app.

So I'm not sure which I should expect? a way to do it without adding the read scope for gphotos? or a way to make if not require verification when you do? Suggestions?

My previous "testing" app (with the required scopes) worked properly, except for the weekly need to go through the OAuth authentication process in a browser.

**Comment by gilesknap** [comment]
Hmm. Sorry it seems that Google has changed the key generation UI again.
I'll have to try and go through to update the doc.

**Comment by wdirksen** [comment]
I'm u and running about 6 weeks on ArchARM. Certificate failure has been frequent lately necessitating a --new-token command 2x in the past month. So this is due to Google changing things on the fly? What is the normal expectation here?

<code>
12-21 16:53:16 WARNING  gphotos-sync 2.14.2+28.g298259a 2020-12-21 16:53:16.641592
12-21 16:53:17 WARNING  Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate is not yet valid (_ssl.c:1123)'))': /$discovery/rest?version=v1 
</code>

**Comment by gilesknap** [comment]
@wdirksen  to my knowledge there has been no change. my own long standing key and that of the project's CI tests are both renewing without issues for over 2 years. Your error looks like it is SSL related rather than token related. 'certificate is not yet valid' makes it sound like a clock issue maybe??

**Comment by wdirksen** [comment]
> @wdirksen to my knowledge there has been no change. my own long standing key and that of the project's CI tests are both renewing without issues for over 2 years. Your error looks like it is SSL related rather than token related. 'certificate is not yet valid' makes it sound like a clock issue maybe??

Thanks @gileknap. System time is in NZST and is correct but machine locale based on US which I'm reading could cause problems with SSL. Working with another new token and hoping it will last awhile.

**Comment by spantaleev** [comment]
I'm running gphotos-sync on a fixed schedule, once per day. I'm hitting this issue as well - about once per week. This makes using gphotos-sync to keep an automated backup of your photos infeasible.

Any way we can debug this? When does it try to renew the token? I can try to collect logs of all runs over the next week.

**Comment by wdirksen** [comment]
I have found the solution for me. You have to 'publish' the app to allow
functioning of repeat tokens. I'm now on week 3 without retrieving new
token since publishing when it used to be 7 days. See newest readme info in
latest git submission.

**Comment by synman** [comment]
publishing my "apps" worked for me too.   I've been running a couple weeks now with no auth token timeouts.

**Comment by bamsnow** [comment]
Does token expiration happen on reboot? I have gphotos_sync running on 20.04, and it appears the only time I have to reauthenticate is on reboot. Is that normal behavior?

**Comment by gilesknap** [comment]
The token should not ever expire if everything is working well. There is a refresh handshake that should always happen when it does expire. The  the refresh token is saved next to your db file so if that gets removed then you will see a new authentication request, otherwise you should not (unless you are having the issues in this thread which are to do with how the initial token got created)

**Comment by bamsnow** [comment]
@gilesknap yes this is what I was referring to, the refresh handshake. I have gphoto-sync running on a UPS-backed headless VM and it would be great to figure out some way to make a handshake request never occur and just have gphoto-sync run forever.

**Comment by gilesknap** [comment]
The handshake I'm referring to is automatic with no human intervention required.

My gphotos sync has run for several years without asking me to log in. But I got my token before google started making it harder. I think the clues to getting it right are in this thread but I have not tried it for years.
________________________________

**Comment by gilesknap** [comment]
I'm hoping that authorization issues are fixed in 3.0 since the auth flow has been updated.

**Comment by gilesknap** [comment]
I'm releasing a new version of gphotos-sync and closing stale issues. Please reopen if you still want me to look at this. Thanks.
